core/integration/module_test.go

```diff
@@ -3104,6 +3104,73 @@ class Test {
 	}
 }
 
+func TestCustomModuleEnumType(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		sdk    string
+		source string
+	}
+	for _, tc := range []testCase{
+		{
+			sdk: "typescript",
+			source: `import { func, object, field, daggerEnum } from "@dagger.io/dagger"
+
+/**
+ * Enum for Status
+ */
+@daggerEnum()
+class Status {
```

Comment:
Question: why this syntax instead of typescript's builtin enum type? https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

I think we should stick to the language semantics if possible.

Reply:
Do we need a decorator here? Can we not take the same approach as in go, where we create typedefs for every referenced type - e.g. so if a function accepts/returns an enum, then we automatically codegen for it.

Maybe that's less inline with the rest of the typescript SDK, and makes for more complications.

Reply:
I mean I wish it worked like that, but I need to refactor a lot of the Typescript internal introspection logic to get that working.

Should I work on that?

Reply:
I think not being able to use a decorator in an enum shouldn't be an argument against using `enum` if that's the most natural thing. That said, I remember a user saying they didn't like TypeScript's enums, but I don't know why.

Reply:
Also, part of me thinks `@function` is the only one we could need. Both for class methods and properties. In Python however, `@object` does change the class, unlike TypeScript. If not that, users would need `@dataclass` anyway.

Reply:
> I think we should stick to the language semantics if possible.  

I strongly agree with this.  That being said, enums in typescript can act fairly weirdly.  The general way, that at least I use, is to just use a type literal as its more natural in Typescript.  But that ends up with the same issue of not being able to add a decorator.

>  That said, I remember a user saying they didn't like TypeScript's enums, but I don't know why.

I left it on a comment in a PR, I'll try to find it.

> Also with Helder we aggreed that explicit declaration is better than implicit, I could add a special case for enum but I don't know, I'm not sure.

The explicit declaration is a little cumbersome in Typescript, as it generally adds a bunch of boiler plate, especially for things like this enum

Reply:
Here was my original comment on enums.  
https://github.com/dagger/dagger/issues/7401#issuecomment-2125999323

Reply:
As I said, not being able to use a decorator shouldn't be an argument against using something else that is more natural. The problem with a type of string literals really is how do you add descriptions to the values?